nightly: add cluster object

petermattis opened the pull request:
Add `cluster` object which makes creating and running commands on a
cluster somewhat more concise.

Add support for running "local" clusters via a `--local` flag.

Release note: None

petermattis committed:
nightly: add cluster object

Add `cluster` object which makes creating and running commands on a
cluster somewhat more concise.

Add support for running "local" clusters via a `--local` flag.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22515)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4qitsQ9qJ-iyuqXjI6:-L4qitsQ9qJ-iyuqXjI7:b-37welq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L28)):*
> ```Go
> )
> 
> type cluster struct {
> ```

This deserves commentary, but I wanted to get feedback on the approach first.

---

*[pkg/nightly/nightly_test.go, line 209 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4qj-1ZT_fPjKFIsOsa:-L4qj-1ZT_fPjKFIsOsb:bn6d8qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L209)):*
> ```Go
> 			c.put(ctx, *workload, "<workload>")
> 			c.start(ctx, 1, 3)
> 			c.ssh(ctx, 4, testCmd)
> ```

This is my attempt to reduce boiler plate. Internally, the various methods call `t.Fatal` on error. The use of `t.Helper` ensures that the fatal errors are tied back to these source lines.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

benesch commented:
Approach LGTM, modulo my one comment below.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4r6QoXOnJ8Z7e6meOP:bf0or8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
> ```Go
> type cluster struct {
> 	name string
> 	t    *testing.T
> ```

We've been burned by embedding `testing.T`s in structs before. If a parent test creates a cluster and a subtest triggers a call to `t.Fatal` on the parent's `t` via the cluster object, things go haywire. Better to pass `t` to every function, annoying as it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4r6rbQmf0rCfeMUQ7U:bpfepi5)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rF1tUiT_tf_zkF4Us:b-amtt5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We've been burned by embedding `testing.T`s in structs before. If a parent test creates a cluster and a subtest triggers a call to `t.Fatal` on the parent's `t` via the cluster object, things go haywire. Better to pass `t` to every function, annoying as it is.
</blockquote></details>

Is it parent and child tests that are problematic, or calling methods on the `testing.T` from goroutines? Do you have a pointer to where we've been burned by this before. I want to understand the problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

danhhz commented:
approach also lgtm

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rC17VyB1Ux4Q8ckn0:bexmmrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We've been burned by embedding `testing.T`s in structs before. If a parent test creates a cluster and a subtest triggers a call to `t.Fatal` on the parent's `t` via the cluster object, things go haywire. Better to pass `t` to every function, annoying as it is.
</blockquote></details>

=benesch

---

*[pkg/nightly/nightly_test.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rCePK0LdokWqVmV_T:-L4rCePK0LdokWqVmV_U:b-zbr0ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L76)):*
> ```Go
> 			return "", err
> 		}
> 		path = gopath + "/src/github.com/cockroachdb/cockroach/" + binary
> ```

filepath.Join?

---

*[pkg/nightly/nightly_test.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDqrLtAQeU1T99l8T:-L4rDqrLtAQeU1T99l8U:bdr4csr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L124)):*
> ```Go
> 			t.Parallel()
> 		} else {
> 			initLocalBinaries(t)
> ```

i'm a little confused why local vs non-local has such a big impact on the behavior of these flags. my vote is unify the behavior (unless i'm missing something?)

---

*[pkg/nightly/nightly_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDR8RtQG2EIjsYYqP:-L4rDR8RtQG2EIjsYYqQ:bx2u8f5) ([raw file](https://github.com/cockroachdb/cockroach/blob/60351e11490e69a87bf10d090fd8c87dfcb3f158/pkg/nightly/nightly_test.go#L161)):*
> ```Go
> 			}
> 
> 			m := monitorWithContext(ctx)
> ```

i assume you'll add this back when polishing

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4rEZk9IsQBgEX0dJOx:bkh1mii)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rFOvnzwD2i3TTCEIN:b-i0j5m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

=benesch
</blockquote></details>

parent and child is the big one here. testutils.SQLRunner is the most recent example. i (and others) constantly got that one wrong before we made every method take a testing.TB

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/nightly_test.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDqrLtAQeU1T99l8T:-L4rFwnQ-s_DETrrSYU6:btu7ztf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L124)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i'm a little confused why local vs non-local has such a big impact on the behavior of these flags. my vote is unify the behavior (unless i'm missing something?)
</blockquote></details>

For the local case I don't want you to have to specify the binary, but locate it via `GOPATH`. For remote that seems unlikely to work. Am I missing something about how we can unify the behavior?

---

*[pkg/nightly/nightly_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDR8RtQG2EIjsYYqP:-L4rG7P7F0JeqhbUYh96:b-dtivae) ([raw file](https://github.com/cockroachdb/cockroach/blob/60351e11490e69a87bf10d090fd8c87dfcb3f158/pkg/nightly/nightly_test.go#L161)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

i assume you'll add this back when polishing
</blockquote></details>

Oops, this was accidentally removed when I rebased. I'll re-add once I get the `cluster/testing.T` thing figured out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/nightly_test.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDqrLtAQeU1T99l8T:-L4rGqUZKuqSg-8yqfWH:bwaf0gr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L124)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

For the local case I don't want you to have to specify the binary, but locate it via `GOPATH`. For remote that seems unlikely to work. Am I missing something about how we can unify the behavior?
</blockquote></details>

IMO using the initLocalBinaries behavior for both is the least confusing. in the non-local mode, if a non-linux binary is found, it should error and point you to the flag to override

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rLCBwkqDEUS9UJpqw:b-85t6ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

parent and child is the big one here. testutils.SQLRunner is the most recent example. i (and others) constantly got that one wrong before we made every method take a testing.TB
</blockquote></details>

Ok, thanks for the pointer. I'll take a look.

---

*[pkg/nightly/nightly_test.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rCePK0LdokWqVmV_T:-L4rKclc-TyVv_IqU1NK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L76)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

filepath.Join?
</blockquote></details>

Done.

---

*[pkg/nightly/nightly_test.go, line 124 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDqrLtAQeU1T99l8T:-L4rKdSKd5FNgTV4pGGH:biafqp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/nightly_test.go#L124)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

IMO using the initLocalBinaries behavior for both is the least confusing. in the non-local mode, if a non-linux binary is found, it should error and point you to the flag to override
</blockquote></details>

Hmm, ok. I suppose we'll always be setting the flags from `nightly-workload.sh`. I wasn't planning to check to see if the binary is built for the appropriate platform. I figure that will work itself out when you try to start the cluster. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rN5HwYNbsOvDRTqK7:b-42ci9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ok, thanks for the pointer. I'll take a look.
</blockquote></details>

So the problematic pattern is something like?

```
func TestParent(t *testing.T) {
  c := newCluster(t)

  t.Run("child", func(t *testing.T) {
    c.doSomething(...)
  })
}
```

That will be all kinds of broken if we allow subtests to run in parallel. Just want to make sure I'm on the same page regarding the problematic pattern.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rOk3HhDOl0T3yr4c3:bi3uat9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

So the problematic pattern is something like?

```
func TestParent(t *testing.T) {
  c := newCluster(t)

  t.Run("child", func(t *testing.T) {
    c.doSomething(...)
  })
}
```

That will be all kinds of broken if we allow subtests to run in parallel. Just want to make sure I'm on the same page regarding the problematic pattern.
</blockquote></details>

yes. that's exactly the issue

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

benesch commented:
Yep, that’s exactly it. If c.doSomething calls t.Fatal, the error produced
is incredibly obtuse, even when the subtests are not parallelized.

On Thu, Feb 8, 2018 at 4:47 PM Peter Mattis <notifications@github.com>
wrote:

> Review status: 0 of 3 files reviewed at latest revision, 6 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *pkg/nightly/cluster_test.go, line 30 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rN5HwYNbsOvDRTqK7:b-42ci9n>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Ok, thanks for the pointer. I'll take a look.
>
> So the problematic pattern is something like?
>
> func TestParent(t *testing.T) {
>   c := newCluster(t)
>
>   t.Run("child", func(t *testing.T) {
>     c.doSomething(...)
>   })
> }
>
> That will be all kinds of broken if we allow subtests to run in parallel.
> Just want to make sure I'm on the same page regarding the problematic
> pattern.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/22515>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-364259045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IJp0p84K5i2wz7dCJKbGhXUpGK9sks5tS2tXgaJpZM4R-1n8>
> .
>

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rPDCNsaQCXulRsJ37:bxogsnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

yes. that's exactly the issue
</blockquote></details>

Passing the `testing.T` as a parameter will help somewhat, but perhaps we want to disallow sharing of a `cluster` between tests completely. I'm trying to think of a way to do that. One option is to pass the `testing.T` as a parameter and assert that it is the same as the `testing.T` used to create the `cluster`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rQqyTAgjL12V-ACJN:btq7g2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Passing the `testing.T` as a parameter will help somewhat, but perhaps we want to disallow sharing of a `cluster` between tests completely. I'm trying to think of a way to do that. One option is to pass the `testing.T` as a parameter and assert that it is the same as the `testing.T` used to create the `cluster`. 
</blockquote></details>

```
func TestLocal(t *testing.T) {
	ctx := context.Background()
	c := newCluster(ctx, t)

	t.Run("child", func(t *testing.T) {
		c.run(ctx, t, "roachprod", "status", c.name)
	})
}
...
=== RUN   TestLocal/child
panic: cluster created by TestLocal, but used by TestLocal/child [recovered]
	panic: cluster created by TestLocal, but used by TestLocal/child
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/nightly/cluster_test.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rSX7wSJ3vKwGKIfsP:bxse53n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

```
func TestLocal(t *testing.T) {
	ctx := context.Background()
	c := newCluster(ctx, t)

	t.Run("child", func(t *testing.T) {
		c.run(ctx, t, "roachprod", "status", c.name)
	})
}
...
=== RUN   TestLocal/child
panic: cluster created by TestLocal, but used by TestLocal/child [recovered]
	panic: cluster created by TestLocal, but used by TestLocal/child
```
</blockquote></details>

Ok, this is ready for another look. See `TestInvalidClusterUsage` below. I think I also need to address the problem with using a `cluster` from multiple goroutines. Currently that won't work very well as the `t.Fatal` call will cause the current goroutine exit, not the main test goroutine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

benesch commented:
It’s not clear to me that it’s unreasonable to share a cluster among
multiple subtests—you could imagine each roachmart region being its own
subtest. And if one subtest’s SSH command fails, you need only fail that
subtest.

That said, happy to do the stricter thing first.

On Thu, Feb 8, 2018 at 5:01 PM Peter Mattis <notifications@github.com>
wrote:

> Review status: 0 of 3 files reviewed at latest revision, 6 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *pkg/nightly/cluster_test.go, line 30 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4r6QoWP9CEvsz_oaYK:-L4rQqyTAgjL12V-ACJN:btq7g2m>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1ababa981c34c5854a94b77cb9e950ee69d64612/pkg/nightly/cluster_test.go#L30>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Passing the testing.T as a parameter will help somewhat, but perhaps we
> want to disallow sharing of a cluster between tests completely. I'm
> trying to think of a way to do that. One option is to pass the testing.T
> as a parameter and assert that it is the same as the testing.T used to
> create the cluster.
>
> func TestLocal(t *testing.T) {
> 	ctx := context.Background()
> 	c := newCluster(ctx, t)
>
> 	t.Run("child", func(t *testing.T) {
> 		c.run(ctx, t, "roachprod", "status", c.name)
> 	})
> }
> ...
> === RUN   TestLocal/child
> panic: cluster created by TestLocal, but used by TestLocal/child [recovered]
> 	panic: cluster created by TestLocal, but used by TestLocal/child
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/22515>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-364263189>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IER68z8tEJ75hdGzRBgebQhronu3ks5tS26zgaJpZM4R-1n8>
> .
>

petermattis commented:
> It’s not clear to me that it’s unreasonable to share a cluster among
multiple subtests—you could imagine each roachmart region being its own
subtest. And if one subtest’s SSH command fails, you need only fail that
subtest.

Not sure exactly what you're suggesting there due to lack of familiarity with roachmart, but it sounds like an abuse of subtests.

benesch commented:
The point is I can see how you might want to use one cluster for multiple
tests. Happy to cross that bridge if we come to it, though.

On Thu, Feb 8, 2018 at 5:12 PM Peter Mattis <notifications@github.com>
wrote:

> It’s not clear to me that it’s unreasonable to share a cluster among
> multiple subtests—you could imagine each roachmart region being its own
> subtest. And if one subtest’s SSH command fails, you need only fail that
> subtest.
>
> Not sure exactly what you're suggesting there due to lack of familiarity
> with roachmart, but it sounds like an abuse of subtests.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-364266099>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15INrIPS8ivQUXLxdGC_CghnXRmegXks5tS3FAgaJpZM4R-1n8>
> .
>

petermattis commented:
Ok, this is ready for another look.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4rec6Cse5Grah_TCK7:bp9rmyd)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/nightly/cluster_test.go, line 97 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4ulflRQ9eVx2Nen-NK:-L4ulflRQ9eVx2Nen-NL:bq5457) ([raw file](https://github.com/cockroachdb/cockroach/blob/7214ddd96158a43b86dc4e28f400b62b725b5685/pkg/nightly/cluster_test.go#L97)):*
> ```Go
> 		}
> 	}
> 	if t.Failed() {
> ```

why is this check (and the one in Start) after the local cluster stuff? I'd expect them to be treated the same

---

*[pkg/nightly/cluster_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4um2PruCh2W4KXWP2e:-L4um2PruCh2W4KXWP2f:b-71eb4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7214ddd96158a43b86dc4e28f400b62b725b5685/pkg/nightly/cluster_test.go#L127)):*
> ```Go
> }
> 
> // Stop cockroach nodes running on a subset of the cluster. See start() for a
> ```

nit: clarify which `start()` this is referring to

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4unTrZhP8ZmyTrHPiM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
TFTR! Last thing to do is to get the monitor stuff hooked up again. Will do so today.

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/nightly/cluster_test.go, line 97 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4ulflRQ9eVx2Nen-NK:-L4unq5bXMLk_7_8bALN:b-qim9ce) ([raw file](https://github.com/cockroachdb/cockroach/blob/7214ddd96158a43b86dc4e28f400b62b725b5685/pkg/nightly/cluster_test.go#L97)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

why is this check (and the one in Start) after the local cluster stuff? I'd expect them to be treated the same
</blockquote></details>

No reason. They can certainly be moved earlier in the function. I've now regularized this.

---

*[pkg/nightly/cluster_test.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4um2PruCh2W4KXWP2e:-L4unxzbOly4o1SwCAN0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7214ddd96158a43b86dc4e28f400b62b725b5685/pkg/nightly/cluster_test.go#L127)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: clarify which `start()` this is referring to
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4uou5G1G5noViBZwto:b-a2n9m7)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/nightly/nightly_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4rDR8RtQG2EIjsYYqP:-L4vOhhYUi2imWjrjCKX:byuj9br) ([raw file](https://github.com/cockroachdb/cockroach/blob/60351e11490e69a87bf10d090fd8c87dfcb3f158/pkg/nightly/nightly_test.go#L161)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Oops, this was accidentally removed when I rebased. I'll re-add once I get the `cluster/testing.T` thing figured out.
</blockquote></details>

I baked the monitoring stuff into `cluster`. Mostly an adaptation of your code, @danhhz. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

But it seems pretty clear to me that we should make roachprod a library soon.

---

Reviewed 1 of 3 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4vQ9U-A7_-VKW2jVtV:b-1um6cj)*
<!-- Sent from Reviewable.io -->

danhhz commented:
still lgtm

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 219 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRLeGc4DGBDBbWiNQ:-L4vRLeGc4DGBDBbWiNR:b-9wh8at) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/cluster_test.go#L219)):*
> ```Go
> 		cmd := exec.CommandContext(ctx, args[0], args[1:]...)
> 		cmd.Stdout = pipeW
> 		cmd.Stderr = c.l.stderr
> ```

my version made a sublogger, which was nice because these stood out in the logs

---

*[pkg/nightly/nightly_test.go, line 221 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vQm5nrkiq1rVHtsmp:-L4vPgPY9ZKtGziqrguS:b-hroisa) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/nightly_test.go#L221)):*
> ```Go
> 			c.Start(ctx, t, 1, 3)
> 
> 			g := &errgroup.Group{}
> ```

probably better to use errgroup.WithContext. the zero value of errgroup.Group doesn't cancel the other goroutines when one of them fails. not relevant here, but if this gets cargo culted (say for TestPartitioning), it won't do the right thing

---

*[pkg/nightly/nightly_test.go, line 276 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRwZVqA1n6N_2eIs9:-L4vRwZVqA1n6N_2eIsA:b-4ltmxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/nightly_test.go#L276)):*
> ```Go
> 		}
> 
> 		m := monitorWithContext(ctx)
> ```

don't forget to port this one to the new thing and delete the old thing

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4vS-EJwEsC6fdCf10i:bh6086p)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> But it seems pretty clear to me that we should make roachprod a library soon.

Yeah, some of the stuff in `cluster` (such as the monitoring) would be simplified, but it isn't too bad exec'ing `roachprod` and gives us discipline in how we extend `roachprod`.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/nightly/cluster_test.go, line 219 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRLeGc4DGBDBbWiNQ:-L4vorSqYeKZFOqGbR4p:b-1dj99l) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/cluster_test.go#L219)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

my version made a sublogger, which was nice because these stood out in the logs
</blockquote></details>

Yeah, I didn't do that because `stdLogger` doesn't work nicely with sub-loggers  (i.e. it creates a file like `std_MONITOR`). Let me add some TODOs. 

---

*[pkg/nightly/nightly_test.go, line 221 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vQm5nrkiq1rVHtsmp:-L4voBUSpdV9B2BvgFo9:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/nightly_test.go#L221)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

probably better to use errgroup.WithContext. the zero value of errgroup.Group doesn't cancel the other goroutines when one of them fails. not relevant here, but if this gets cargo culted (say for TestPartitioning), it won't do the right thing
</blockquote></details>

Good point. Done.

---

*[pkg/nightly/nightly_test.go, line 276 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRwZVqA1n6N_2eIs9:-L4voIs0kLA-5PcVj4yq:b-aqccc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/nightly_test.go#L276)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

don't forget to port this one to the new thing and delete the old thing
</blockquote></details>

I was leaving this alone assuming you guys have outstanding changes to it. Do you? Mind if I address in a follow-on PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4voVnaUZQsVQySGgwi:b8r2t8u)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 1 of 3 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/cluster_test.go, line 219 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRLeGc4DGBDBbWiNQ:-L4vsCUOIzWi1-ooK0Au:b182rpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/cluster_test.go#L219)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yeah, I didn't do that because `stdLogger` doesn't work nicely with sub-loggers  (i.e. it creates a file like `std_MONITOR`). Let me add some TODOs. 
</blockquote></details>

aha. ack

---

*[pkg/nightly/nightly_test.go, line 276 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRwZVqA1n6N_2eIs9:-L4vs-vPtM81SIZlGWyj:b-e6jlht) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/nightly_test.go#L276)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I was leaving this alone assuming you guys have outstanding changes to it. Do you? Mind if I address in a follow-on PR?
</blockquote></details>

to my knowledge, neither of us do, but follow-on is perfectly fine

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis committed:
nightly: make childLogger work with stdLogger

Refactor `logger` to allow `logger.childLogger` to work with
`stdLogger`. `stdLogger` is path-thru to stdout/stderr, but when a child
logger is created we wrap `prefixWriters` around those writers.

Release note: None

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/nightly/cluster_test.go, line 219 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L4vRLeGc4DGBDBbWiNQ:-L4w5zfWeKQzfYWA4KKv:bt8ja40) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3d56f29f9ca0d7b74a111cf0a3f86ecb9d9e89/pkg/nightly/cluster_test.go#L219)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

aha. ack
</blockquote></details>

I added a 2nd commit which refactors `logger` so that `childLogger` works well with `stdLogger`. This ended up replacing `dualLogger` with `io.MultiWriter` and `prefixWriter`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

petermattis committed:
nightly: update TestRoachmart to use cluster

Release note: None

petermattis committed:
nightly: fix TestSingleDC

Add support for monitoring a subset of nodes. `TestSingleDC` starts
cockroach no nodes 1-3 and then runs the `workload` on node 4.

Release note: None

petermattis committed:
nightly: remove some detritus

Silence a lint warning about usage of `testing.T.Fatal`.

Release note: None

petermattis committed:
nightly: code movement

Move initialization code from `nightly_test.go` to `main_test.go`.

Release note: None

petermattis commented:
I ended up adding a few more commits to this PR. One updates `TestRoachmart`. Another fixes `TestSingleDC` by allowing monitoring of a subset of nodes. And then a few cleanup commits (moving code to `main_test.go` and `cluster_test.go` and removing detritus).

There seems to be a bug somewhere in the tear down code. Sometimes if you ctrl-c a test the local cluster will not be destroyed. Probably also happens for remote tests. I don't think tracking that down should hold up this PR, just letting you know that I'm aware of it. 

---

Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L4wZNgKG_x0eLNtzUpu:byyqx1s)*
<!-- Sent from Reviewable.io -->

petermattis committed:
nightly: rework cluster monitoring

The previous version of monitoring had a variety of races between
workers finishing, the monitoring command exiting and cancelation which
caused flakiness in the `TestClusterMonitor` tests.

Release note: None

petermattis commented:
> There seems to be a bug somewhere in the tear down code. Sometimes if you ctrl-c a test the local cluster will not be destroyed. Probably also happens for remote tests. I don't think tracking that down should hold up this PR, just letting you know that I'm aware of it.

Tracked this down to a problem with destroying multi-node local clusters. See https://github.com/cockroachdb/roachprod/pull/58/commits/ca2ba062f19f47d5bcbdc464078d6c57a4c43520

petermattis committed:
nightly: simplify monitor node specification

`roachprod monitor` already allows monitoring a subset of nodes, so no
need to replicate this functionality in `monitor.wait()`.

Release note: None

petermattis committed:
nightly: fix checkTestTimeout

Release note: None

petermattis commented:
PS I'm looking for a final LGTM on the most recent changes. I pinky promise I won't extend the scope of this PR again.

tschottdorf commented:
This looks fine for now (based on a coarse review), though I'm left with a
bit of an uneasy feeling because I don't want to see the `acceptance`
`Cluster` story repeated (where we wanted ready-made clusters that worked
well for simple tests, but could never satisfy ones that needed more
control).

Is the plan that with a perf baseline nightly test in place, we can take a
step back and prioritize teasing out modular components which are useful in a
variety of stability setups? That sounds reasonable as it'll take some time,
though in the meantime I'll have concerns about this `*cluster* thing gaining
weight it may never lose.

PS: I'm not a fan of the `*testing.T` buy-in, though I suppose that's a
matter of personal preference.

PPS: the ergonomics of `*cluster` look fine, though I'm not sure they're on
the right receiver.

---

Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L58H0mNonQiwBHgBUcl:b-b57xso)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> This looks fine for now (based on a coarse review), though I'm left with a
> bit of an uneasy feeling because I don't want to see the acceptance
> Cluster story repeated (where we wanted ready-made clusters that worked
> well for simple tests, but could never satisfy ones that needed more
> control).

Yeah, I think there are unresolved questions about how this will work for more complex tests. My plan to address that is to start working on some of the more complex tests soon. For example, a long running chaos test and a mixed-version test seem like necessities. I'm curious if you have any more specific concerns here. The `cluster` object allows creating local, remote and geo clusters. You can stop and start individual nodes, wipe and restart the cluster, etc.

> Is the plan that with a perf baseline nightly test in place, we can take a
step back and prioritize teasing out modular components which are useful in a
variety of stability setups? That sounds reasonable as it'll take some time,
though in the meantime I'll have concerns about this `cluster thing gaining
weight it may never lose.

As mentioned above, I'd like to get a few additional tests in place before teasing out reusable components. In particular, chaos and multi-version tests are likely to expose deficiencies in the current approach. 

> PS: I'm not a fan of the *testing.T buy-in, though I suppose that's a
matter of personal preference.

We're discussing whether this nightly test stuff should use `go test` at all. The alternative would be some sort of `roachtest` binary. Is this what you're referring to, or are you talking about something else?

> PPS: the ergonomics of *cluster look fine, though I'm not sure they're on
the right receiver.

Can you elaborate on this? Which receiver would be better?

---

Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L59PPfRuJ93r3C3q12C:bje9rf1)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> I look forward to seeing how this works with chaos and multi-version

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/log_test.go, line 103 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L59q0pVFAUQ0lL1cwO3:-L59q0pVFAUQ0lL1cwO4:b-8lttfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/640da5d809774bf07c58d1dce0709f1e0f49a314/pkg/nightly/log_test.go#L103)):*
> ```Go
> 	// newline. No newline and the data is buffered. We don't have a signal for
> 	// when the end of data is reached, which means we won't output any trailing
> 	// data if it isn't terminated with a newline.
> ```

should we make this implement WriteCloser, so there isn't this surprise? not sure it's worth the complexity but figured i'd put the suggestion out there

---

*[pkg/nightly/main_test.go, line 61 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L59qRf31GMGSbokr0yA:-L59qRf31GMGSbokr0yB:bdja840) ([raw file](https://github.com/cockroachdb/cockroach/blob/640da5d809774bf07c58d1dce0709f1e0f49a314/pkg/nightly/main_test.go#L61)):*
> ```Go
> 	}
> 
> 	// For "local" clusters we have to find the binary to run and translate it to
> ```

We do this for remote cluster now, too, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L59qr0KOWVffPW6bq8T:bintnoz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/nightly/log_test.go, line 103 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L59q0pVFAUQ0lL1cwO3:-L59rW9SfAmFPqY6SF_9:b-o3g141) ([raw file](https://github.com/cockroachdb/cockroach/blob/640da5d809774bf07c58d1dce0709f1e0f49a314/pkg/nightly/log_test.go#L103)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

should we make this implement WriteCloser, so there isn't this surprise? not sure it's worth the complexity but figured i'd put the suggestion out there
</blockquote></details>

That has repercussions: `logger.stdout` would need to implement `WriteCloser` which seems awkward.

---

*[pkg/nightly/main_test.go, line 61 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L59qRf31GMGSbokr0yA:-L59ruszzaR1GRc1kkSH:b-zh2yfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/640da5d809774bf07c58d1dce0709f1e0f49a314/pkg/nightly/main_test.go#L61)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

We do this for remote cluster now, too, right?
</blockquote></details>

Yep. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/nightly/log_test.go, line 103 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-L59q0pVFAUQ0lL1cwO3:-L59shL6cJGB_2YDIxBF:b-8ly7mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/640da5d809774bf07c58d1dce0709f1e0f49a314/pkg/nightly/log_test.go#L103)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That has repercussions: `logger.stdout` would need to implement `WriteCloser` which seems awkward.
</blockquote></details>

Gotcha :thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok, let me try to be a little more concrete with this criticism (and praise).

> As mentioned above, I'd like to get a few additional tests in place before teasing out reusable components. In particular, chaos and multi-version tests are likely to expose deficiencies in the current approach.

We have both chaos and mixed-version tests in `acceptance` already. My take-away from that bit of history is that it's a slippery slope to make an opinionated framework for simple tests first and to then hope that it'll grow organically to accommodate complicated ones that need more control.

I derive from this my preference to keep tests explicit, to initially bury very little complexity in the general test harness, and to establish clear interfaces early. What I was hoping to see is a basic `roachprod` wrapper that provisions resources and provides some access to them, and tests running explicit commands or using reusable components such as chaos, monitoring, etc (which sometimes have to know about each other). Then, it's fine to and make sugar that reduces boilerplate for certain classes of tests, but now the sugar is the foundational building block, and this worries me due to the limitations and precedent it sets.

This PR in itself doesn't have enough code to set matters in stone, but I don't exactly see your momentum going the way I had hoped and so I'm worried I'm going to have a hard time warming up to the follow-ups as well. I thought I had put some effort in to prevent that from happening; I want this stuff to be a lot better than the `acceptance` tests. I would be interested to know whether whether I just didn't do a good job stressing these points in https://github.com/cockroachdb/roachprod/pull/27. Perhaps what I'm trying to achieve isn't adequate or premature, but it hasn't seemed that way to me.

On the plus side, it's not quite as bad as `acceptance` just yet. I like that tests get to deploy their own binaries because that leaves tests in control of that. Also, the way the simple tests pan out is nice (mod having the cockroach incantation buried which is quite problematic) and is in line with how I expected the sugar for them to work (so my criticism is specifically on what `*cluster` is composed of, not how it's currently used from the existing tests).

> We're discussing whether this nightly test stuff should use go test at all. The alternative would be some sort of roachtest binary. Is this what you're referring to, or are you talking about something else?

Yeah, something like that. I'm quite flexible on the exact form taken.

> Can you elaborate on this? Which receiver would be better?

One that isn' `*cluster` (or, a version of `*cluster` that carries less cockroach-specific logic). Basically just my earlier points.

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5AHTykgYzzE86QIAyh:b-h3nh8t)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Thanks for the detailed response. I'm not at all settled on the approach for this new testing infrastructure and still in the learning/exploration phase. Your viewpoint is very appreciated.

> We have both chaos and mixed-version tests in acceptance already. My take-away from that bit of history is that it's a slippery slope to make an opinionated framework for simple tests first and to then hope that it'll grow organically to accommodate complicated ones that need more control.

Agreed. My plan is to get a nice sampling of different tests before trying to add any more opionation or convenience. 

> I derive from this my preference to keep tests explicit, to initially bury very little complexity in the general test harness, and to establish clear interfaces early. What I was hoping to see is a basic roachprod wrapper that provisions resources and provides some access to them, and tests running explicit commands or using reusable components such as chaos, monitoring, etc (which sometimes have to know about each other). Then, it's fine to and make sugar that reduces boilerplate for certain classes of tests, but now the sugar is the foundational building block, and this worries me due to the limitations and precedent it sets.

Heh, I read the first few sentences in this paragraph and I think it describes what we have in the `cluster` object. Yet the last sentence indicates you're seeing something different. Can you elaborate on what sugar you're noticing is a foundational building block. From my perspective, `cluster` is a thin wrapper on top of `roachprod`. The only functionality of significant complexity is the monitoring.

> Yeah, something like that. I'm quite flexible on the exact form taken.

https://github.com/cockroachdb/cockroach/pull/22593

> > Can you elaborate on this? Which receiver would be better?
>
> One that isn' *cluster (or, a version of *cluster that carries less cockroach-specific logic). Basically just my earlier points.

What cockroach specific logic is concerning you? Yeah, `cluster` knows how to start a cockroach cluster, but really it is just calling out to `roachprod` to do that. Would you want each test to be specifying the exact cockroach command line to run on each node?

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5Bddn1Cqjv1q_wjEQx:b-e97gsf)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Heh, I read the first few sentences in this paragraph and I think it describes what we have in the cluster object. Yet the last sentence indicates you're seeing something different. Can you elaborate on what sugar you're noticing is a foundational building block. From my perspective, cluster is a thin wrapper on top of roachprod. The only functionality of significant complexity is the monitoring.

Yes, sorry. Much of my criticism is really aimed at `roachprod` or its being called. I don't think we ought to be calling `roachprod start` as it's [extremely rigid](https://github.com/cockroachdb/roachprod/blob/master/cockroach.go#L14). It's worth thinking about this early before it's too unattractive to back out of. This is really just repeating [earlier pseudocode](https://github.com/cockroachdb/roachprod/pull/27/files#diff-a63e31c84f3b8ec0e29f56865e31f78fR62) but here is what I hoped to see:

```go
c := newCluster(...) // more like newNodes

// Omitted: distribute binaries

c.Run(c.All(), "cockroach start --isnecure --store {{STore 0}} --join {{Addr 0}}")

// Set any perf-related cluster settings, etc (or rather, call some helper that does it)

monitor := SimpleSQLMonitor(pgUrls, `SELECT 1`)
go monitor.Run()
```

An upshot of this is that code reads at a verbosity similar to a shell session. I think this a good thing as it gives more intuition on what is happening, which directly affects how easy it is to prototype and extend code. For perf numbers, it also makes it easier to find out what's actually running.
Looking up what `roachprod start` does is a lot harder and you'll be scared to change anything as the code is shared. But, more importantly, I have an easier time imagining how chaos and other intricate tests are strung together with primitives. I can even imagine how most tests can be run against a local cluster with minimal changes (thanks to the interpolation, which has default values provided in `c` and those can differ based on local/remote), though I think it wise to postpone that.

Similarly, for the monitoring, I'd prefer establishing a clear interface early. A monitor is some check that is run periodically and needs to know very little except for what to check -- in the simple cases sufficient for perf testing `SELECT 1` is already enough and so a handful of PGUrls and an interval is all it needs to see. There's no need to plant this component into `*cluster`, quite the opposite - that makes it hard to add things like chaos and risks weaving them into the monolith as well.

I'd be happy to contribute a toy test for this (as sort of a reality check, too), but I assume you're working full steam and I'd just get in your way?

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5BrJCr85igBEOenT2p:b-ieqok2)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
PS: sorry about being somewhat of a downer. For the simple tests, this looks really good. I'm just scared that things will turn ugly when you leave that realm.

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5CkLBaipKbhBEVQwgv:bg343ho)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> Yes, sorry. Much of my criticism is really aimed at roachprod or its being called. I don't think we ought to be calling roachprod start as it's extremely rigid.

`roachprod start` might be less rigid than you think. `roachprod start foo:<nodes>` can start arbitrary nodes in the cluster. `roachprod start -a <args>` can supply arbitrary arguments. Also note that a test isn't restricted to using `roachprod start` as it could start the nodes via `roachprod ssh` (a.k.a `cluster.Run`). The one deficiency here is that `roachprod start` has special handling for node 1, but I'd like to get rid of that by using `cockroach init`.

For chaos, I was imagining that a test would `roachprod start` and then select a random node and `roachprod stop <cluster>:<node>; sleep ...; roachprod start <cluster>:<node>`. 

> Similarly, for the monitoring, I'd prefer establishing a clear interface early. A monitor is some check that is run periodically and needs to know very little except for what to check -- in the simple cases sufficient for perf testing SELECT 1 is already enough and so a handful of PGUrls and an interval is all it needs to see. There's no need to plant this component into *cluster, quite the opposite - that makes it hard to add things like chaos and risks weaving them into the monolith as well.

The existing `monitor` is a process monitor. I'm imagining the tests will need more sophisticated monitors as the tests themselves become more sophisticated. I'm not understanding your last point about `planting this component into *cluster`: `monitor` is distinct from `cluster`. Monitoring a chaos is definitely going to be a challenge as you need to "expect" certain node deaths. I haven't wrapped my head around exactly what I want to do there. Perhaps there is a separate monitor per node. Having a single stream of monitoring events as we did in `acceptance` was fragile. 

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5EQ7LUqcugtwj0XNMS:b-cjdwh6)*
<!-- Sent from Reviewable.io -->

petermattis commented:
>  I can even imagine how most tests can be run against a local cluster with minimal changes (thanks to the interpolation, which has default values provided in c and those can differ based on local/remote), though I think it wise to postpone that.

Not sure if you noticed, but running tests against local clusters pretty much just works right now. With the `roachtest` PR, you can do `roachtest run -l kv0/nodes=0`. The one caveat is that many tests will want to specify different flags and concurrency when running locally.

---

Review status: 0 of 4 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5ETDL1p8MLHG7uLQyd:byw3od0)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I did notice that, and I mentioned this for parity with this PR. I think it
was premature to introduce that mode so early.

On Tue, Feb 13, 2018, 9:04 AM Peter Mattis <notifications@github.com> wrote:

> I can even imagine how most tests can be run against a local cluster with
> minimal changes (thanks to the interpolation, which has default values
> provided in c and those can differ based on local/remote), though I think
> it wise to postpone that.
>
> Not sure if you noticed, but running tests against local clusters pretty
> much just works right now. With the roachtest PR, you can do roachtest
> run -l kv0/nodes=0. The one caveat is that many tests will want to
> specify different flags and concurrency when running locally.
> ------------------------------
>
> Review status: 0 of 4 files reviewed at latest revision, 13 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5ETDL1p8MLHG7uLQyd:byw3od0>*
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-365275553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135P1hE7aUtBFqHOfVCbY9umLN6xcOks5tUZZOgaJpZM4R-1n8>
> .
>


-- 

-- Tobias

benesch commented:
I actually don’t think we introduced that flag soon enough. Iterating on
tests without the local flag is incredibly painful.

On Tue, Feb 13, 2018 at 10:35 PM Tobias Schottdorf <notifications@github.com>
wrote:

> I did notice that, and I mentioned this for parity with this PR. I think it
> was premature to introduce that mode so early.
>
> On Tue, Feb 13, 2018, 9:04 AM Peter Mattis <notifications@github.com>
> wrote:
>
> > I can even imagine how most tests can be run against a local cluster with
> > minimal changes (thanks to the interpolation, which has default values
> > provided in c and those can differ based on local/remote), though I think
> > it wise to postpone that.
> >
> > Not sure if you noticed, but running tests against local clusters pretty
> > much just works right now. With the roachtest PR, you can do roachtest
> > run -l kv0/nodes=0. The one caveat is that many tests will want to
> > specify different flags and concurrency when running locally.
> > ------------------------------
> >
> > Review status: 0 of 4 files reviewed at latest revision, 13 unresolved
> > discussions, all commit checks successful.
> > ------------------------------
> >
> > *Comments from Reviewable
> > <
> https://reviewable.io:443/reviews/cockroachdb/cockroach/22515#-:-L5ETDL1p8MLHG7uLQyd:byw3od0
> >*
> >
> > —
> > You are receiving this because you commented.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-365275553
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AE135P1hE7aUtBFqHOfVCbY9umLN6xcOks5tUZZOgaJpZM4R-1n8
> >
> > .
> >
>
>
> --
>
> -- Tobias
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/22515#issuecomment-365485720>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IJ8OxsECyE21rPZBmx2Wb8d-911jks5tUlRjgaJpZM4R-1n8>
> .
>

tschottdorf commented:
@benesch this implies that only the flag allows fast iteration, which isn't correct. The flag makes it harder to iterate on the infrastructure itself.
